### PR TITLE
[stable-3_3_0] pkp/pkp-lib#6663 Use recorded MIME type if available

### PR DESCRIPTION
**Pull request by Vitaliy-1**

### CODE REVIEW

File: classes/services/PKPFileService.inc.php [hunk 1]
```diff
@@ -117,22 +117,35 @@ public function delete($id) {
...
+	 * @param string|int $pathOrFileId The path to the file or file ID
 	 * @param string $filename Filename to give to the downloaded file
 	 * @param boolean $inline Whether to stream the file to the browser
 	 */
-	public function download($path, $filename, $inline = false) {
+	public function download($pathOrFileId, $filename, $inline = false) {
+
+		$dispatcher = Application::get()->getRequest()->getDispatcher();
+
+		if (is_integer($pathOrFileId)) {
```
Comment (NateWr) [inline classes/services/PKPFileService.inc.php:128]: Use `is_int` instead. Just for consistency.

File: classes/services/PKPFileService.inc.php [hunk 2]
```diff
@@ -117,22 +117,35 @@ public function delete($id) {
...
+			$path = $pathOrFileId;
+		}
 
 		if (!$this->fs->has($path)) {
-			Application::get()->getRequest()->getDispatcher()->handle404();
+			$dispatcher->handle404();
 		}
 
-		if (HookRegistry::call('File::download', [$path, &$filename, $inline])) {
+		if (HookRegistry::call('File::download', [$file ?? $path, &$filename, $inline])) {
```
Comment (NateWr) [inline classes/services/PKPFileService.inc.php:143]: Don't modify the hook. We don't want the hooks to change between stable versions.

File: classes/services/PKPFileService.inc.php [hunk 3]
```diff
@@ -117,22 +117,35 @@ public function delete($id) {
...
 			return;
 		}
 
 		// Stream the file to the end user.
-		$mimetype = $this->fs->getMimetype($path) ?? 'application/octet-stream';
+		$mimetype = $file->mimetype ?? $this->fs->getMimetype($path) ?? 'application/octet-stream';
```
Comment (NateWr) [inline classes/services/PKPFileService.inc.php:148]: Did you test this with a path? I think calling `$file->mimetype` willl cause an error if only the path was called...

**Review by NateWr — CHANGES_REQUESTED** [review]

File: classes/services/PKPFileService.inc.php (hunk 1 above)
Comment (Vitaliy-1) [reply classes/services/PKPFileService.inc.php]: fixed

File: classes/services/PKPFileService.inc.php (hunk 2 above)
Comment (Vitaliy-1) [reply classes/services/PKPFileService.inc.php]: Thanks! Fixed

File: classes/services/PKPFileService.inc.php (hunk 3 above)
Comment (Vitaliy-1) [reply classes/services/PKPFileService.inc.php:148]: Yes, tested it. It first will check if `$file` and it's mimetype property is initialized, before null check; shouldn't cause a problem.

Comment (NateWr) [reply classes/services/PKPFileService.inc.php:148]: Thanks, you're right. I didn't believe it but the null coalescing operator works on object properties even when the object isn't defined. :tada:

**Pull request merged**